service/api/service.go

```diff
@@ -157,10 +158,24 @@ func (s *serviceImpl) ApiV1WorkflowWaitForStateCompletion(
 	defer func() { log.CapturePanic(recover(), s.logger, &retError) }()
 
 	var workflowId string
-	if req.WaitForKey != nil {
-		workflowId = service.IwfSystemConstPrefix + req.WorkflowId + "_" + *req.StateId + "_" + *req.WaitForKey
-	} else {
-		workflowId = service.IwfSystemConstPrefix + req.WorkflowId + "_" + *req.StateExecutionId
+
+	waitForOn := env.GetSharedConfig().Api.WaitForStateCompletionMigration.WaitForOn
+
+	if waitForOn == "old" {
+		workflowId = utils.GetWorkflowIdForWaitForStateExecution(req.WorkflowId, *req.StateExecutionId, *req.WaitForKey, *req.StateId)
+	} else { // waitForOn == "new"
+		var parentId string
+		if s.client.GetBackendType() == service.BackendTypeTemporal { // Temporal
+			response, err := s.client.DescribeWorkflowExecution(ctx, req.GetWorkflowId(), "", nil)
+			if err != nil {
+				return nil, s.handleError(err, WorkflowWaitForStateCompletionApiPath, req.WorkflowId)
+			}
+			parentId = response.FirstRunId
+		} else { // Cadence
+			parentId = req.WorkflowId
+		}
+
+		workflowId = utils.GetWorkflowIdForWaitForStateExecution(parentId, *req.StateExecutionId, *req.WaitForKey, *req.StateId)
```

Comment:
`*req.WaitForKey` will assume that field is always present (otherwise it will be a panic -- access to nil value). I think only StateExecutionId is required. So we probably need to pass pointer into the method and let the method do a nil check

Reply:
Fixed